pallets/did/src/lib.rs

```diff
@@ -28,93 +28,345 @@ mod tests;
 
 	/// Weight information for extrinsics in this pallet.
 	type WeightInfo: WeightInfo;
+}
+
+/// Trait representing a public key under the control of a DID subject.
+pub trait DIDPublicKey {
+	/// Returns the key method description as in the [DID specification](https://w3c.github.io/did-spec-registries/#verification-method-types).
+	fn get_did_key_description(&self) -> &'static str;
+}
+
+/// Enum representing the types of verification keys a DID can control.
+#[derive(Clone, Decode, Debug, Encode, Eq, Ord, PartialEq, PartialOrd)]
+pub enum PublicVerificationKey {
+	/// An Ed25519 public key.
+	Ed25519([u8; 32]),
+	/// A Sr25519 public key.
+	Sr25519([u8; 32]),
+}
+
+impl PublicVerificationKey {
+	/**
+	Given a payload and a signature, the specific public verification key will return either
+	a SignatureVerificationResult if the signature is not properly formed, or a boolean indicating
+	the result of the verification.
+	 */
+	fn verify_signature(&self, payload: &Payload, signature: &Signature) -> SignatureVerificationResult {
+		// Discard all invalid signatures by comparing them with the expected length of the specific verification key.
+		ensure!(signature.len() == self.get_expected_signature_size(), SignatureError::InvalidSignatureFormat);
+		
+		// Signature::try_from expects a &[u8], so we cast the vec (of the right size) to the expected type.
+		let signature = signature as &[u8];
+		match self {
+			PublicVerificationKey::Ed25519(raw_key) => {
+				// Try to re-create a Signature value or throw an error if raw value is invalid.
+				let ed25519_sig = ed25519::Signature::try_from(
+					signature
+				).map_err(|_| SignatureError::InvalidSignatureFormat)?;
+				// Re-create a Public value from the raw value of the key.
+				let signer = ed25519::Public(raw_key.clone());
+				// Returns the result of the signature verification.
+				Ok(ed25519_sig.verify(payload as &[u8], &signer))
+			}
+			// Follows same process as above, but using a Sr25519 instead.
+			PublicVerificationKey::Sr25519(raw_key) => {
+				let sr25519_sig = sr25519::Signature::try_from(
+					signature
+				).map_err(|_| SignatureError::InvalidSignatureFormat)?;
+				let signer = sr25519::Public(raw_key.clone());
+				Ok(sr25519_sig.verify(payload as &[u8], &signer))
+			}
+		}
+    }
+
+	/// Returns the expected signature length (of bytes) for the given verification key type.
+    fn get_expected_signature_size(&self) -> usize {
+        match self {
+			PublicVerificationKey::Ed25519(_) | PublicVerificationKey::Sr25519(_) => {
+				64
+			}
+		}
+    }
+}
+
+impl DIDPublicKey for PublicVerificationKey {
+    fn get_did_key_description(&self) -> &'static str {
+		match self {
+			PublicVerificationKey::Ed25519(_) => "Ed25519VerificationKey2018",		// https://w3c.github.io/did-spec-registries/#ed25519verificationkey2018
+			PublicVerificationKey::Sr25519(_) => "Sr25519VerificationKey2020"		// Not yet defined in the DID specification.
+		}
```

Comment:
Just had an idea how you can have your traits and dynamic dispatch. [If you return a reference it, can also be dynamic](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8bcf4161d364c32714fc408cc89a13a9).

Not sure if that helps, but it might be worth a try?